src/sinks/clickhouse.rs

```diff
@@ -0,0 +1,254 @@
+use crate::{
+    buffers::Acker,
+    event::Event,
+    sinks::util::{
+        http::{HttpRetryLogic, HttpService},
+        retries::FixedRetryPolicy,
+        BatchServiceSink, Buffer, Compression, SinkExt,
+    },
+    topology::config::{DataType, SinkConfig},
+};
+use futures::{Future, Sink};
+use http::{Method, Uri};
+use hyper::{Body, Client, Request};
+use hyper_tls::HttpsConnector;
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+use tower::ServiceBuilder;
+
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct ClickhouseConfig {
+    pub host: String,
+    pub table: String,
+    pub batch_size: Option<usize>,
+    pub batch_timeout: Option<u64>,
+    pub compression: Option<Compression>,
+
+    // Tower Request based configuration
+    pub request_in_flight_limit: Option<usize>,
+    pub request_timeout_secs: Option<u64>,
+    pub request_rate_limit_duration_secs: Option<u64>,
+    pub request_rate_limit_num: Option<u64>,
+    pub request_retry_attempts: Option<usize>,
+    pub request_retry_backoff_secs: Option<u64>,
+}
+
+#[typetag::serde(name = "clickhouse")]
+impl SinkConfig for ClickhouseConfig {
+    fn build(&self, acker: Acker) -> Result<(super::RouterSink, super::Healthcheck), String> {
+        let sink = clickhouse(self.clone(), acker);
+        let healtcheck = healthcheck(self.host.clone());
+
+        Ok((sink, healtcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+fn clickhouse(config: ClickhouseConfig, acker: Acker) -> super::RouterSink {
+    let host = config.host.clone();
+    let table = config.table.clone();
+    let gzip = match config.compression.unwrap_or(Compression::Gzip) {
+        Compression::None => false,
+        Compression::Gzip => true,
+    };
+
+    let batch_size = config.batch_size.unwrap_or(bytesize::mib(10u64) as usize);
+    let batch_timeout = config.batch_timeout.unwrap_or(1);
+
+    let timeout = config.request_timeout_secs.unwrap_or(60);
+    let in_flight_limit = config.request_in_flight_limit.unwrap_or(5);
+    let rate_limit_duration = config.request_rate_limit_duration_secs.unwrap_or(1);
+    let rate_limit_num = config.request_rate_limit_num.unwrap_or(5);
+    let retry_attempts = config.request_retry_attempts.unwrap_or(usize::max_value());
+    let retry_backoff_secs = config.request_retry_backoff_secs.unwrap_or(1);
+
+    let policy = FixedRetryPolicy::new(
+        retry_attempts,
+        Duration::from_secs(retry_backoff_secs),
+        HttpRetryLogic,
+    );
+
+    let http_service = HttpService::new(move |body: Vec<u8>| {
+        let uri = format!(
+            "{}/?query=INSERT%20INTO%20{}%20FORMAT%20JSONEachRow",
```

Comment:
1. It might be better to put the table name inside double quotes or backticks, so that names that can't be used as unquoted identifiers still can be used as quoted identifiers. See https://clickhouse.yandex/docs/en/query_language/syntax/#syntax-identifiers. For example, it is possible to call table `my.table_name`, then insert into it like `INSERT INTO "my.table_name" ...`. Without the quotes ClickHouse would understand `my.table_name` as table `table_name` in database `my`. The table name inside a quoted identifier should also be escaped (for double quotes all `"` should be replaced by `\"`).

2. As quoted identifiers can contain spaces, the query string should be percent-encoded after substitution of the table name. Doing percent-encoding in a function of the can also improve readability of the code by keeping spaces instead of `%20` in the source code.

Reply:
I've added this https://github.com/timberio/vector/pull/693/commits/c6497d66dd73452d1a661c9ed161071bb97ff895 let me know if this works with what you were mentioning above.